File: Serenity.Script.Core/Q/Q.Services.cs
using System;
using System.Html;
using System.Net;
using System.Runtime.CompilerServices;
using jQueryApi;
using System.Collections.Generic;

namespace Serenity
{
    public static partial class Q
    {
        [InlineCode("Q.serviceCall({options})")]
        public static extern XmlHttpRequest ServiceCall(ServiceCallOptions options);

        [ScriptName("serviceCall"), IncludeGenericArguments(false)]
        public static jQueryXmlHttpRequest ServiceCall<TResponse>(ServiceCallOptions<TResponse> options)
            where TResponse: ServiceResponse
        {
            Action<ServiceResponse> handleError = delegate(ServiceResponse response) 
            {
                if (Q.Config.NotLoggedInHandler != null &&
                    response != null &&
                    response.Error != null &&
                    response.Error.Code == "NotLoggedIn" &&
                    Q.Config.NotLoggedInHandler(options.As<ServiceCallOptions>(), response))
                                return;

                    if (options.OnError != null)
                        options.OnError(response);
                    else
                        Q.ErrorHandling.ShowServiceError(response.Error);
            };

            options = jQuery.ExtendObject(new ServiceCallOptions<TResponse>
            {
                DataType = "json",
                ContentType = "application/json",
                Type = "POST",
                Cache = false,
                BlockUI = true,
                Url = options.Service != null && !options.Service.StartsWith('~') && !options.Service.StartsWith('/') ?
                    Q.ResolveUrl("~/services/" + options.Service) : Q.ResolveUrl(options.Service),
                Data = Q.ToJSON(options.Request),
                Success = (data, textStatus, request) =>
                {
                    var response = data.As<TResponse>();
                    try
                    {
                        if (response.Error == null)
                        {
                            if (options.OnSuccess != null)
                                options.OnSuccess(response);
                        }
                        else
                        {
                        }
                    }
                    finally
                    {
                        if (options.BlockUI)
                            Q.BlockUndo();

                        if (options.OnCleanup != null)
                            options.OnCleanup();
                    }
                },
                Error = (xhr, status, ev) =>
                {
                    try
                    {
                        if (xhr.Status == 403)
                        {
                            string l = null;
                            try { l = xhr.GetResponseHeader("Location"); }
                            catch { l = null; }
                            if (l != null)
                            {
                                Window.Top.Location.Href = l;
                                return;
                            }
                        }

                        if ((xhr.GetResponseHeader("content-type") ?? "").ToLower().IndexOf("application/json") >= 0)
                        {
                            var json = jQuery.ParseJsonData<ServiceResponse>(xhr.ResponseText);
                            if (json != null && json.Error != null)
                            {
                                handleError(json);
                                return;
                            }
                        }

                        var html = xhr.ResponseText;

                        Q.Externals.IFrameDialog(new { html = html });
                    }
                    finally
                    {
                        if (options.BlockUI)
                            Q.BlockUndo();

                        if (options.OnCleanup != null)
                            options.OnCleanup();
                    }
                }
            }, options);

            if (options.BlockUI)
                Q.BlockUI();

            return jQuery.Ajax(options.As<jQueryAjaxOptions>());
        }

        [IncludeGenericArguments(false)]
        public static void ServiceRequest<TResponse>(string service, ServiceRequest request, Action<TResponse> onSuccess, ServiceCallOptions options = null)
            where TResponse: ServiceResponse
        {
            ServiceCall(jQuery.ExtendObject(new ServiceCallOptions<TResponse>
            {
                Service = service,
                Request = request,
                OnSuccess = onSuccess
            }, options.As<ServiceCallOptions<TResponse>>()));
        }
    }
}

File: Serenity.Web/Scripts/serenity/CoreLib/Q/Q.Services.ts
/// <reference path="Q.BlockUI.ts" />
/// <reference path="Q.Config.ts" />
/// <reference path="Q.ErrorHandling.ts" />
/// <reference path="Q.Url.ts" />
/// <reference path="../Services/ServiceResponse.ts" />

namespace Q {
    export interface ServiceOptions<TResponse extends Serenity.ServiceResponse> extends JQueryAjaxSettings {
        request?: any;
        service?: string;
        blockUI?: boolean;
        onError?(response: TResponse): void;
        onSuccess?(response: TResponse): void;
        onCleanup?(): void;
    }

    export function serviceCall<TResponse>(options: Q.ServiceOptions<TResponse>) {
        let handleError = function (response: any) {
            if (Config.notLoggedInHandler != null &&
                response &&
                response.Error &&
                response.Error.Code == 'NotLoggedIn' &&
                Config.notLoggedInHandler(options, response)) {
                return;
            }

            if (options.onError != null) {
                options.onError(response);
                return;
            }

            ErrorHandling.showServiceError(response.Error);
        };

        var url = options.service;
        if (url && url.length && url.charAt(0) != '~' && url.charAt(0) != '/' && url.indexOf('://') < 0)
            url = resolveUrl("~/services/") + url;

        options = $.extend<Q.ServiceOptions<TResponse>>({
            dataType: 'json',
            contentType: 'application/json',
            type: 'POST',
            cache: false,
            blockUI: true,
            url: url,
            data: $.toJSON(options.request),
            success: function (data, textStatus, request) {
                try {
                    if (!data.Error && options.onSuccess) {
                        options.onSuccess(data);
                    }
                }
                finally {
                    if (options.blockUI) {
                        blockUndo();
                    }
                    options.onCleanup && options.onCleanup();
                }
            },
            error: function (xhr, status, ev) {
                try {
                    if (xhr.status === 403) {
                        var l: any = null;
                        try {
                            l = xhr.getResponseHeader('Location');
                        }
                        catch ($t1) {
                            l = null;
                        }
                        if (l) {
                            window.top.location.href = l;
                            return;
                        }
                    }
                    if ((xhr.getResponseHeader('content-type') || '')
                        .toLowerCase().indexOf('application/json') >= 0) {
                        var json = $.parseJSON(xhr.responseText);
                        if (json && json.Error) {
                            handleError(json);
                            return;
                        }
                    }
                    var html = xhr.responseText;
                    iframeDialog({ html: html });
                }
                finally {
                    if (options.blockUI) {
                        blockUndo();
                    }
                    options.onCleanup && options.onCleanup();
                }
            }
        }, options);

        if (options.blockUI) {
            blockUI(null);
        }

        return $.ajax(options);
    }

    export function serviceRequest<TResponse>(service: string, request?: any,
        onSuccess?: (response: TResponse) => void, options?: Q.ServiceOptions<TResponse>) {
        return serviceCall($.extend<Q.ServiceOptions<TResponse>>({
            service: service,
            request: request,
            onSuccess: onSuccess
        }, options));
    }

    export function setEquality(request: Serenity.ListRequest, field: string, value: any) {
        if (request.EqualityFilter == null) {
            request.EqualityFilter = {};
        }
        request.EqualityFilter[field] = value;
    }
}

namespace Serenity {
    export type ServiceOptions<TResponse extends Serenity.ServiceResponse> = Q.ServiceOptions<TResponse>;
}